### PR TITLE
Make EditOrder command give a warning if the date is in the past

### LINKED ISSUE #43: Refactor Order Date to throw specific error categories

### LINKED ISSUE #124: Make EditOrder command give a warning on duplicate order
Make `editOrder` command give a warning to the user, telling that there is a duplicate warning if:
* There is an existing order with the same item name and date
* Both the edited order and duplicate order have pending status

### LINKED ISSUE #125: Make AddOrder command give a warning if the given date is before the current date
Currently users can freely add an order with a delivery date that is in the past.
We can make the command throw a warning to the user if the delivery date is in the past since it is likely a mistake made by the user.

### LINKED ISSUE #126: Make EditOrder command give a warning if the date is in the past
Currently `editOrder` command allows users to edit the date of a command to be in the past.
Make the command such that it gives a warning if that is the case since it may be a mistake on the users part.

### PR DESCRIPTION
**Pull request by kaixuan477**
Refactor Order Date to throw specific error categories, make AddOrder and EditOrder commands give a warning if date is in the past.

Close #126 
Close #125 
Close #43

### CODE REVIEW

File: src/main/java/seedu/sellsavvy/model/order/Date.java [hunk 1]
```diff
@@ -33,24 +39,50 @@ public class Date {
...
+        } catch (ParseException e) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Returns {@code Date} warning if order date has passed.
+     */
+    public String hasDateElapsed() {
```
Comment (awdse22) [inline src/main/java/seedu/sellsavvy/model/order/Date.java:72]: Given the method name is `hasDateElapsed`, wouldn't it make more sense if the method returns a boolean instead?
In EditOrderCommand, you can add `MESSAGE_INVALID_DATE` to the `feedbackToUser` variable if the date has already elapsed.

File: src/main/java/seedu/sellsavvy/model/order/Order.java [hunk 2]
```diff
@@ -60,6 +60,13 @@ public boolean isSameOrder(Order otherOrder) {
                 && otherOrder.getItem().equals(getItem());
     }
 
+    /**
+     * Returns {@code Date} warning if order date has passed.
+     */
+    public String hasDateElapsed() {
```
Comment (awdse22) [inline src/main/java/seedu/sellsavvy/model/order/Order.java:66]: Same applies here for the earlier comment regarding returning boolean.

File: src/test/java/seedu/sellsavvy/model/order/DateTest.java [hunk 3]
```diff
@@ -20,24 +22,52 @@ public void constructor_invalidDate_throwsIllegalArgumentException() {
...
+    public void isValidDateRegex() {
         // null date value
-        assertThrows(NullPointerException.class, () -> Date.isValidDate(null));
+        assertThrows(NullPointerException.class, () -> Date.isValidDateRegex(null));
+
+        // invalid date regex
+        assertFalse(Date.isValidDateRegex("")); // empty string
+        assertFalse(Date.isValidDateRegex(" ")); // spaces only
+        assertFalse(Date.isValidDateRegex("date")); // not a date
+        assertFalse(Date.isValidDateRegex("2-02-2002")); // invalid day format
```
Comment (awdse22) [inline src/test/java/seedu/sellsavvy/model/order/DateTest.java:33]: Can this be renamed to `single digit day`?

Comment (awdse22) [inline src/test/java/seedu/sellsavvy/model/order/DateTest.java:33]: On top of single digit days, can a single digit month be included as well?
e.g. 02-2-2002

File: src/test/java/seedu/sellsavvy/model/order/DateTest.java [hunk 4]
```diff
@@ -20,24 +22,52 @@ public void constructor_invalidDate_throwsIllegalArgumentException() {
...
+        assertFalse(Date.isValidDateRegex("02/02/2002")); // non-hyphen non-numeric characters
+        assertFalse(Date.isValidDateRegex("02-22-2002")); // invalid month
+        assertFalse(Date.isValidDateRegex("02-02-2222")); // not within 21st century
```
Comment (awdse22) [inline src/test/java/seedu/sellsavvy/model/order/DateTest.java:36]: Can there be another test case for 20th century or earlier? That EP isn't covered here.

**Review by awdse22 — COMMENTED** [review]
LGTM, just a few nitpicks.

File: src/test/java/seedu/sellsavvy/model/order/DateTest.java (hunk 3 above)
Comment (kaixuan477) [reply src/test/java/seedu/sellsavvy/model/order/DateTest.java:33]: Both single digit days and month are currently not supported. You can reformat to allow for formatting to d-m-yy but thats just QoL.

File: src/main/java/seedu/sellsavvy/model/order/Date.java (hunk 1 above)
Comment (kaixuan477) [reply src/main/java/seedu/sellsavvy/model/order/Date.java]: Updated in commit.

File: src/main/java/seedu/sellsavvy/model/order/Order.java (hunk 2 above)
Comment (kaixuan477) [reply src/main/java/seedu/sellsavvy/model/order/Order.java]: Updated in commit.

File: src/test/java/seedu/sellsavvy/model/order/DateTest.java (hunk 3 above)
Comment (kaixuan477) [reply src/test/java/seedu/sellsavvy/model/order/DateTest.java]: Updated in commit.

Comment (kaixuan477) [reply src/test/java/seedu/sellsavvy/model/order/DateTest.java]: Updated in commit.

File: src/test/java/seedu/sellsavvy/model/order/DateTest.java (hunk 4 above)
Comment (kaixuan477) [reply src/test/java/seedu/sellsavvy/model/order/DateTest.java]: Updated in commit.

**Review by awdse22 — APPROVED** [review]
LGTM

**Pull request merged**